pkg/threescale/backend/manager.go

```diff
@@ -0,0 +1,367 @@
+type Application struct {
+	LastResponse threescale.AuthorizeResult
+	LimitCounter LimitCounter
+	//ReportWithValues
+	UnlimitedHits map[string]int
+}
+
+// Limit captures the current state of the rate limit for a particular time period
+type Limit struct {
+	current int
+	max     int
+}
+
+// LimitCounter keeps a count of limits for a given period
+type LimitCounter map[string]map[api.Period]*Limit
+
+// Request Transactions must not be nil and must not be empty
+func (b *Backend) Authorize(request threescale.Request) (threescale.AuthorizeResult, error) {
+	app, _, err := b.getApplicationFromRequest(request, request.Transactions[0])
+	if err != nil {
+		return nil, fmt.Errorf("unable to process request - %s", err.Error())
+	}
+
+	affectedMetrics := b.computeAffectedMetrics(app.LastResponse.GetHierarchy(), request.Transactions[0].Metrics)
+	_, isAuthorized := b.isAuthorized(app, affectedMetrics)
+	return AuthResponse{success: isAuthorized}, nil
+}
+
+// AuthRep
+// Request Transactions must not be nil and must not be empty
+func (b *Backend) AuthRep(request threescale.Request) (threescale.AuthorizeResult, error) {
+	app, cacheKey, err := b.getApplicationFromRequest(request, request.Transactions[0])
+	if err != nil {
+		return nil, fmt.Errorf("unable to process request - %s", err.Error())
+	}
+
+	affectedMetrics := b.computeAffectedMetrics(app.LastResponse.GetHierarchy(), request.Transactions[0].Metrics)
+	clonedApplication, isAuthorized := b.isAuthorized(app, affectedMetrics)
+	if isAuthorized {
+		b.cache.Set(cacheKey, clonedApplication)
+	}
+
+	return AuthResponse{success: isAuthorized}, nil
+}
+
+// Request Transactions must not be nil and must not be empty
+func (b *Backend) Report(request threescale.Request) (threescale.ReportResult, error) {
+	var hierarchy api.Hierarchy
+
+	processed := make(chan Application, len(request.Transactions))
+	processingErr := make(chan error)
+
+	respondWith := ReportResponse{
+		accepted:  true,
+		processed: processed,
+		error:     processingErr,
+	}
+
+	go func() {
+		for _, transaction := range request.Transactions {
+			// figure out the cachekey for this particular transaction
+			cacheKey := b.generateCacheKeyFromRequest(request, transaction)
+
+			//check to see if its in the cache
+			app, ok := b.cache.Get(cacheKey)
+			if !ok {
+
+				if hierarchy == nil {
+					// we only need to do this once because the hierarchy is at a service and not application level
+					app, err := b.handleCacheMiss(request, cacheKey)
+					processingErr <- err
+					if err != nil {
+						//todo handle error here - we need the hierarchy, do we just abort?
+						break
+					}
+					hierarchy = app.LastResponse.GetHierarchy()
+				} else {
+					// we can work on this locally
+					app = b.newApplication()
+				}
+			}
+			affectedMetrics := b.computeAffectedMetrics(hierarchy, transaction.Metrics)
+			appWithUpdatedState := b.localReport(cacheKey, app, affectedMetrics)
+			processed <- appWithUpdatedState
+		}
```

Comment:
are you happy with providing a buffered channel here? I'm not sure how useful it will be, but i anticipate we might want some mechanism for blocking etc?

Reply:
I'd want to see the fixed logic first :)

Reply:
will remove